vote/src/vote_account.rs

```diff
@@ -303,8 +320,8 @@ impl TryFrom<AccountSharedData> for VoteAccountInner {
             return Err(Error::InvalidOwner(*account.owner()));
         }
         Ok(Self {
+            vote_state: VoteState::deserialize(account.data()).map_err(Error::InstructionError)?,
```

Comment:
@behzadnouri brings up a great point that this deserialization could potentially fail in one of the call-sites (Stakes cache I feel fine with but have to dig into rewards code more carefully). I agree with him that it would be best to have this PR store the `Result` and follow up with another PR that we can carefully review all the call-sites of the `TryFrom` in.

Reply:
If you decide to not split this up (honestly fine with me), I'll do another pass tomorrow after another more careful review. I do think the `OnceLock` approach caused us to introduce the `Result` in the first place so I can see why you coupled that change into this PR.

Reply:
Main concern I had during review was around deserializing `VoteAccounts` (done when deserializing bank stakes or epoch stakes) because after this PR we would now reject snapshots that contain vote accounts that aren't deserializable. This is fine as long as only deserializable vote accounts were ever stored in the stakes cache / epoch stakes. In the current code it's not possible to store a vote account that's not deserializable because the stakes cache wouldn't insert it. But let's say though that in the past we _did_ store a vote account which was not deserializable, it looks like it would have only been removed by `handle_invalid_keys` (called downstream of `update_rewards_with_thread_pool`) if there was a stake account delegated to it because `_load_vote_and_stake_accounts` looks for any vote keys for stake delegations that correspond to invalid vote accounts. But if there are no stake delegations to an invalid vote account in the stakes cache, it doesn't look like we would ever remove it.

In order to merge this I think we should probably try loading snapshots with this branch on each cluster to see if there are some ancient invalid vote accounts in there. Does that make sense?

Reply:
Yep this makes sense. As discussed I've added a custom deserializer that discards invalid vote accounts https://github.com/anza-xyz/agave/pull/2659/commits/356d779e35b0a6bc8598d55309d4e7b9c10eb862

For extra peace of mind I've already checked that there are no invalid accounts in mnb, and I'm going to test the other clusters too.